extract_from_hg.py

```diff
@@ -195,15 +196,110 @@ def commitCset(cset):
+    line = re.sub(r"""
+     \ *         # Match any number of spaces
+     \[*         # Match none or more of `[`, don't worry about `]` - some commit messages do `[r=smith]`
+     rs*         # Match r or rs
+     \=.*        # Match = and the rest of the line.
+    """, "", line, 0, re.VERBOSE)
+
+    # Now change any ': ' to dashes - some people use "Bug 123456: ..."
+    line = re.sub(r": ", " - ", line)
+
+    # Next change any '123456-' to add a space either side of the dash.
+    # Some people do 'Bug 123456-...'.
+    line = re.sub(r"\([0-9]]\)-", r"\1 - ", line)
+
+    # Now replace any commas at the end of the line with dots.
+    line = re.sub(r",$", ".", line)
+
+    # Finally insert '- ' at start of line for changelog formatting.
+    return "- " + line
+
+def writeChangeLog(gitRepo, headGitCommit):
+    inFile = open(CHANGELOG_FILE, "r")
+    oldLines = inFile.readlines()
+    inFile.close()
+
+    outFile = open(CHANGELOG_FILE, "w")
+    foundTBD = False
+    foundFirstDashes = False
+    outBuffer = []
+
+    # Find where to insert the new lines. We hunt down to the first TBD followed
+    # by "---" and then look for the next one. Note that 'i' gets used in the for
+    # statement lower down to finish writing the file.
+    for i in xrange(len(oldLines)):
+        line = oldLines[i]
+        strippedLine = line.rstrip('\n')
+
+        if foundTBD and foundFirstDashes:
+            if line.startswith("---"):
+                # Now we've found the dashes, print the buffer and adjust the index
+                # so that we're ready for later.
+                for bufferLine in outBuffer[:-2]:
+                    outFile.write(bufferLine)
+
+                i = i - 2
+                break
+
+            outBuffer.append(line)
+
+        else:
+            if strippedLine == "TBD":
+                foundTBD = True
+            elif foundTBD and line.startswith("---"):
+                foundFirstDashes = True
+
+            outFile.write(line)
+
+    # Now get the git log entries and add them to the file. If this fails, we
+    # still finish writing the CHANGELOG, so as to not leave the repo in a totally
+    # bad state.
+    try:
+        proc = subprocess.Popen(
+            ['git', 'log',
+             headGitCommit + ".." + gitRepo.head.object.hexsha,
+             "--decorate=no",
+             "--reverse",
+             "--format=format:%s"],
+            stdout=subprocess.PIPE)
+
+        while True:
+            line = proc.stdout.readline()
+            if line == '':
+                break;
+
+            # We shouldn't ever hit this, but just in case...
+            if line.startswith('update latest merged cset file and CHANGELOG'):
+                continue
+
+            outFile.write(formatCommitMessageLine(line) + "\n")
+    except Exception, e:
+        print >> sys.stderr, "Running git log failed: ", e
+
+    # Finally output the rest of the changelog.
+    for j in xrange(i, len(oldLines)):
```

Comment:
For the sake of readability, how about saving off the value of i at the end of the for loop with a more descriptive variable name and using that here?